### DISCUSSION TITLE
Programmatic way to stop a particular DAG Run

### DISCUSSION THREAD

**Discussion by nikhil1699** [apache/airflow#14357] [Q&A] [answered]
We have multiple dags running on a kubernetes pod, And we would like to know if it is possible to Stop one of the running DAG with a dag_id and run_id.
There are some ways to get around this on UI but we want a programmatic or CLI way to achieve the same.

[https://stackoverflow.com/questions/49039386/how-do-i-stop-an-airflow-dag/54205945](url)

**Comment by turbaszek** [comment, marked as answer]
The link does not work, this seems to be correct one:
https://stackoverflow.com/questions/49039386/how-do-i-stop-an-airflow-dag

**Reply by turbaszek** [reply]
What exactly you want to achieve? To you want to stop a DAG and all running tasks but not remove the DAG from Airflow?

**Reply by nikhil1699** [reply]
Thanks for responding :)

Yes exactly. I basically want to achieve STOP and then Start later, But it is fine for us if an already running task keeps running until it finishes, as long as no new task starts execution for that DAG after STOP

**Reply by nikhil1699** [reply]
The link does not work, this seems to be correct one:
https://stackoverflow.com/questions/49039386/how-do-i-stop-an-airflow-dag 
Yes sorry for inconvenience, But that link couldn't solve our problem. As it shows a UI way.

**Reply by turbaszek** [reply]
@nikhil1699 I think you may try "stopping" the running DagRun by deleting it or setting it's state to `failed`. So basically I would suggest checking the code behind the "UI way" and doing the same from script 😉  Or use API:
https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/delete_dag_run

Find a DagRun:
https://github.com/apache/airflow/blob/4c35955bb93ec8276d3b9b3c5994c4be6f96f907/airflow/models/dagrun.py#L233-L235

Change it state:
https://github.com/apache/airflow/blob/4c35955bb93ec8276d3b9b3c5994c4be6f96f907/airflow/models/dagrun.py#L152

This won't stop running task as that would require sending a SIGTERM to executor processes.

**Reply by turbaszek** [reply]
@nikhil1699 did it answer your question?

**Reply by nikhil1699** [reply]
@turbaszek  Thanks for the help. What I am observing was that when I am setting the state to failed, using set_state, it does so in object level, when i use print to check state before and after set_state, it changes the state there, But the DAG doesn't actually stop run.
I think i will have to commit changes to DB as well.
But yes this does solve our problem.

**Reply by turbaszek** [reply]
> I think i will have to commit changes to DB as well.

That's right. You may use this context manager to create session:
https://github.com/apache/airflow/blob/4c35955bb93ec8276d3b9b3c5994c4be6f96f907/airflow/utils/session.py#L27-L28
